src/umode.rs

```diff
@@ -101,6 +265,26 @@ impl<'active> Drop for ActiveUmode<'active> {
 impl<'active> ActiveUmode<'active> {
     /// Run `umode` until completion or error.
     pub fn run(&mut self) -> Result<(), Error> {
-        Ok(())
+        self.run_to_exit();
+        match Trap::from_scause(self.arch.trap_csrs.scause).unwrap() {
+            Trap::Exception(UserEnvCall) => {
+                // Exit on ecall.
+                println!("U-mode ecall!");
+                self.arch.print();
+                Ok(())
+            }
+            _ => Err(Error::UnexpectedTrap),
+        }
+    }
+
+    /// Run until it exits
+    fn run_to_exit(&mut self) {
+        unsafe {
+            // Safe to run umode code as it only touches memory assigned to it through umode mappings.
+            _run_umode(&mut *self.arch as *mut UmodeCpuArchState);
+        }
```

Comment:
Where are you setting up `sstatus`? At minimum, you'll need to make sure `sstatus.SPP` is set to 1 before you `sret`.

Reply:
My understanding is that I shouldn't set it to one?

> The SPP bit indicates the privilege level at which a hart was executing before entering supervisor
mode. When a trap is taken, SPP is set to 0 if the trap originated from user mode, or 1 otherwise.
When an SRET instruction (see Section 3.2.2) is executed to return from the trap handler, the
privilege level is set to user mode if the SPP bit is 0, or supervisor mode if the SPP bit is 1; SPP
is then set to 0.

Reply:
As discussed offline -- you are absolutely right :) I got that backwards.

Reply:
Hm. Okay. The rational I would guess is to be able to have a different `sstatus` when running umode and generic supervisor code. I can see why in XS and FS these could be needed in the future.

Reply:
Set and restored `sstatus`, just to be extremely sure that we're not entering with the wrong `SPP` bit.